### ISSUE TITLE
[MdePkg/Protocols]: New PCI Express Platform Protocol, EFI encoding for the PCI Express features and its attributes (Bugzilla Bug 1954)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2#9749]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 1954
Date: 2019-07-02T06:41:49+00:00
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;
To:   Ashraf Javeed &lt;<ashraf.javeed>&gt;
CC:   ashraf.javeed, felixp, nobody, @niruiyu, @vincent-j-zimmer

Blocker for:  #2194, #2313, #2499, #2500
Last updated: 2020-04-22T15:08:02+00:00

**Comment by tianocore-issues** [comment]
## Comment 8325
Date: 2019-07-02 06:41:49 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

- Industry Specification: ---
- Target OS: ---
- Bugzilla Assignee(s): Ashraf Javeed &lt;<ashraf.javeed>&gt;

Created attachment 360
UEFI ECR Title: Add new interface to retrieve device-specific platform policy in the EFI_PCI_PLATFORM_PROTOCOL &amp; EFI_PCI_OVERRIDE_PROTOCOL, and introduce EFI encodings for the other PCI features and i

The present PI specification version 1.7 defined EFI_PCI_PLATFORM_PROTOCOL / EFI_PCI_OVERRIDE_PROTOCOL does not support retrieving the PCI device-specific platform policies for override in the PCI Bus driver and PCI Root Bridge Resource allocation drivers.

This feature request proposes to add new interface to the PCI Platform protocol and its alias PCI Override Protocol which the PCI Bus driver or PCI Root Bridge Resource allocation driver, can use in future, to override certain PCIe-complaint features, like Maximum Payload Size (MPS), Maximum Read Request Size (MRRS), Extended Tag, etc.

This feature request also proposes the EFI encoding for the following PCIe-complaint features defined in the PCI Express Base Specification version 4 Revision 1; :-
1.	Maximum Payload Size (MPS)
2.	Maximum Read Request Size (MRRS)
3.	Extended Tag 
4.	Relax Order Enable
5.	No Snoop Enable
6.	ASPM support
7.	Common Clock Configuration
8.	Extended SYNC
9.	Atomic Op
10.	LTR Enable
11.	PTM support

Attaching the ECR document with the details of this proposal.

The existing PCI Platform Protocol / PCI Override Protocol has to have new GUID to support this new interface, with the name change as defined in the ECR document.

Attachment: [BZ-360-EFI_PCI_PLATFORM_POLICY Protocol ECR.docx](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-360-EFI_PCI_PLATFORM_POLICY%20Protocol%20ECR.docx)

**Comment by tianocore-issues** [comment]
## Comment 8363
Date: 2019-07-03 05:01:10 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

Assigning to self.

**Comment by tianocore-issues** [comment]
## Comment 8367
Date: 2019-07-04 10:40:25 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

This new feature changes are sent to Edk2 Development community  for review.
Post &quot;Reviewed-By&quot; it shall be moved for Edk2-staging.

Additional EFI encoding for PCIe feature Completion Timeout(CTO) value and CTO detection disable is made in the patch.

**Comment by tianocore-issues** [comment]
## Comment 8648
Date: 2019-07-31 00:21:19 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

Created attachment 369
UEFI ECR Title: Add new interface to retrieve device-specific platform policy in the EFI_PCI_PLATFORM_PROTOCOL &amp; EFI_PCI_OVERRIDE_PROTOCOL, and introduce EFI encodings for the other PCI features and i

latest revision of the ECR draft.
Note that there is no proposal definition change in this document; just the description corrections were made.

Attachment: [BZ-369-EFI_PCI_PLATFORM_POLICY Protocol ECR.docx](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-369-EFI_PCI_PLATFORM_POLICY%20Protocol%20ECR.docx)

**Comment by tianocore-issues** [comment]
## Comment 9668
Date: 2019-09-16 02:03:33 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

Changes are in edk2-staging branch:-
https://github.com/tianocore/edk2-staging/tree/UEFI_PCI_ENHANCE-2

**Comment by tianocore-issues** [comment]
## Comment 10254
Date: 2019-10-31 14:15:23 +0000
From: Felix &lt;<felixp>&gt;

1. GetDevicePolicy() takes PCI device handle as a parameter.
Since the function is used during PCI enumeration, PCI device handle may not exist. It&apos;s better to model the function after EFI_PCI_HOST_BRIDGE_RESOURCE_ALLOCATION_PROTOCOL.PreprocessController()
that takes in root bridge handle and PCI device address.

2. It is my understanding that it is suggested to replace existing PCI Platform and PCI Override protocols with the new protocols.
A more gentle approach towards backward compatibility is suggested.
Two options are possible:
  a) - Add the new protocols
     - Keep the old protocols, but mark them as deprecated
     - Clarify that platform can optionally install new pair of protocols or old pair of protocols, but not both.
  b) Do not change existing protocols. Factor out new functionality into a new pair of protocols:
   EFI_PCI_E_PLATFORM_PROTOCOL and EFI_PCI_E_OVERRIDE_PROTOCOL

**Comment by tianocore-issues** [comment]
## Comment 10311
Date: 2019-11-04 05:35:12 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

My response to the feedback above:-
1. The new GetDevicePolicy() interface takes the device handle since the implementation plan is to configure the new PCI features after the installation of the PCI IO Protocol by the PCI Bus driver; this way the current PCI enumeration flow will not be intruded for configuring the other PCI features.
(Secondly, the PCI Platform Protocol legacy interface GetPciRom() also takes the device handle for retrieving the Option ROM images for the PCI Bus driver, this phase is after the PCI IO protocols are installed).
The PCI address can be obtained from the EFI handle by locating the PCI IO Protocol instance.

2. Please review the ECR attached above. (REF: https://github.com/tianocore/edk2-staging/commit/144d2d3d0a4306266de12a521d617c2cf975bec9).
The legacy interfaces of the PCI Platform Protocol is carried to this new definition to maintain the backward compatibility. Please consider this as the new definition of the existing PCI Platform Protocol (or PCI Override Protocol) to introduce new interface; 
  (A). the proposal to the UEFI PI Spec forum is to have the legacy PCI Platform Protocol definition captured in its addendum marked for deprecation of old definition in near future;  
  (B). the platform can choose to publish any one of the legacy PCI Platform/Override Protocol (as long as it is present in PI Spec.), or it can decide to use the new PCI features configuration from the PCI Bus driver itself and publish any one of the new PCI Platform Protocols.
  (C). my code changes proposal of the PCI Bus driver shall make sure to check   for both the legacy PCI protocols as well as this new protocol definitions. Refer to the following feature change request records to support the enhancement of the PCI Bus driver for the new PCI features configuration:-
  REF: https://bugzilla.tianocore.org/show_bug.cgi?id=2194
  REF: https://bugzilla.tianocore.org/show_bug.cgi?id=2313

**Comment by tianocore-issues** [comment]
## Comment 10312
Date: 2019-11-04 05:36:39 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

My response to the feedback above:-
1. The new GetDevicePolicy() interface takes the device handle since the implementation plan is to configure the new PCI features after the installation of the PCI IO Protocol by the PCI Bus driver; this way the current PCI enumeration flow will not be intruded for configuring the other PCI features.
(Secondly, the PCI Platform Protocol legacy interface GetPciRom() also takes the device handle for retrieving the Option ROM images for the PCI Bus driver, this phase is after the PCI IO protocols are installed).
The PCI address can be obtained from the EFI handle by locating the PCI IO Protocol instance.

2. Please review the ECR attached above. (REF: https://github.com/tianocore/edk2-staging/commit/144d2d3d0a4306266de12a521d617c2cf975bec9).
The legacy interfaces of the PCI Platform Protocol is carried to this new definition to maintain the backward compatibility. Please consider this as the new definition of the existing PCI Platform Protocol (or PCI Override Protocol) to introduce new interface; 
  (A). the proposal to the UEFI PI Spec forum is to have the legacy PCI Platform Protocol definition captured in its addendum marked for deprecation of old definition in near future;  
  (B). the platform can choose to publish any one of the legacy PCI Platform/Override Protocol (as long as it is present in PI Spec.), or it can decide to use the new PCI features configuration from the PCI Bus driver itself and publish any one of the new PCI Platform Protocols.
  (C). my code changes proposal of the PCI Bus driver shall make sure to check   for both the legacy PCI protocols as well as this new protocol definitions. Refer to the following feature change request records to support the enhancement of the PCI Bus driver for the new PCI features configuration:-
  REF: https://bugzilla.tianocore.org/show_bug.cgi?id=2194
  REF: https://bugzilla.tianocore.org/show_bug.cgi?id=2313

**Comment by tianocore-issues** [comment]
## Comment 10365
Date: 2019-11-07 16:51:46 +0000
From: Felix &lt;<felixp>&gt;

Ashraf,

I&apos;m OK with #2. The next step is update of the official proposal in the UEFI mantis system in accordance with your feedback.

I still have concerns regarding #1. You are right. There is already precedence of using PCI handle in PciPlatform protocol. I actually have concerns about that too, but I guess it&apos;s too late to fix that.
I don&apos;t like the fact that an instance of PCI I/O is produced before device if fully configured. For example, a driver that have registered callback on PCI I/O installation would get control prior to the PCI bus driver processing of GetDevicePolicy(). If such driver modifies PCI-E registers, it may break subsequent processing by the PCI bus driver. If such driver tries to save PCI-E configuration to act on it later on, it may end up with an incorrect data.

**Comment by tianocore-issues** [comment]
## Comment 10374
Date: 2019-11-08 10:42:39 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

Felix,
Thanks for your inputs. Sure, I shall follow-up on updating the UEFI mantis for this ECR.

Regarding your concern about publishing EFI handles for the PCI devices before its configuration of the PCI features whose policy attributes are defined with the new interface GetDevicePolicy(), are based on the following two reasons:-
(1) these PCI features are not required to be configured early. It can be also configured late, like in OS.
(2) it was suggested by our internal teams to configure after the Option ROM images are loaded; so that any vendor specific workarounds can be applied to its devices by its own images.

If you check my update on 2194; there is PCD defined &quot;PcdOtherPciFeatures&quot; which the platform firmware can use to separate this PCI feature configuration by the EDK2 PCI Bus driver. I am afraid the platform firmware has the overhead to decide whether the EDK2 PCI Bus driver can configure any particular feature from the list whose policy attributes are defined; or it can choose to stay with its own callback to override any particular feature based on installation of the PCI IO Protocol for that device.

Let me know if you still have concerns about this.
Thanks

**Comment by tianocore-issues** [comment]
## Comment 10407
Date: 2019-11-12 09:22:02 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

Created attachment 428
New interface to PCI Platform Protocol and EFI encodings to PCI features and its attributes

Added backward compatibility description

Attachment: [BZ-428-New interface to PCI Platform Protocol and EFI encodings to PCI features and its attributes.docx](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-428-New%20interface%20to%20PCI%20Platform%20Protocol%20and%20EFI%20encodings%20to%20PCI%20features%20and%20its%20attributes.docx)

**Comment by tianocore-issues** [comment]
## Comment 10432
Date: 2019-11-14 16:30:16 +0000
From: Felix &lt;<felixp>&gt;

Ashraf,

I checked again with our PCI experts.
We still have concerns.
With our proposal PCI bus driver implementer may choose to initialize PCI-E features earlier or later, whereas with your proposal there is no flexibility. PCI-E features must be initialized after PCI I/O installation in the second pass over PCI hierarchy.
There may be a reason to do PCI-E initialization early.
From our experience not configuring certain PCI-E capabilities may affect system&apos;s performance and sometimes system stability.
For example, changing Common Clock settings requires link retraining and MRRS and MPLS of the Downstream devices may change. Incorrect clock selection may degrade system performance or even make the system unusable.

As far as &quot;vendor specific workarounds&quot;, I don&apos;t think your argument is correct. I agree that PCI bus driver loads OpROM images. However, nothing is happening at this point. According to UEFI specification, UEFI driver should not touch hardware in the entry point. Device programming will happen much later, when DriverBinding-&gt;Start function is called. At this point PCI bus driver is out of the scope(it&apos;s Start function has completed).

Finally, our PCI developers are still questioning the approach with changing existing interface. Change of existing interface introduces certain level in inconvenience (legacy code would have to be updated to switch from  PciPlatform to PciPlatform2). This inconvenience has to be justified, by an certain advantage that outweighs the inconvenience. It is not clear to us what the advantage is. As I mentioned in comment 5, there a way to introduce the functionality without changing existing protocols.

**Comment by tianocore-issues** [comment]
## Comment 10454
Date: 2019-11-15 08:46:07 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

(In reply to Felix from comment #11)
&gt; Ashraf,
&gt; 
&gt; I checked again with our PCI experts.
&gt; We still have concerns.
&gt; With our proposal PCI bus driver implementer may choose to initialize PCI-E
&gt; features earlier or later, whereas with your proposal there is no
&gt; flexibility. PCI-E features must be initialized after PCI I/O installation
&gt; in the second pass over PCI hierarchy.
&gt; There may be a reason to do PCI-E initialization early.

I want to implement the PCI Bus driver to initialize this PCIe features, and when I look at the PCI Bus driver flow sequence, I can broadly classify as follows:-
(1) PCI Enumeration
    (a) PCI bus allocation phase
    (b) PCI resource allocation phase (MMIO/IO)
    (c) PCI callbacks (during the above a and b) into the platform firmware through spec defined interfaces
(2) PCI device registration
    (a) installation of the PCI IO Protocols
    (b) PCI option ROM processing and start of the PCI bridge devices
How early can I do this?
The least I have to wait is till phase 1.a. that has to be complete because without the PCI bus allocation, the EP devices cannot be accessed. As per PCIe Base specification I did not find any PCIe features that need to be configured before 1.b., as there is no absolute dependence on the PCI resource allocation for those, hence it can be configured later, that is after the PCI enumeration phase. Can you help me identify from the PCIe Base specification what would be the PCIe-specific features that would need the device-specific platform policy to be configured which the EDK2 PCI Bus driver has to do before phase 1.a.?
Note that I am discounting the following feature set that is related to pre-link configuration; my prime objective is to configure the PCIe features common for both the parent PCIe Root Port and its child EP devices, with regard to performance and stability. If it is pre-link related, like defined below, than it is only for the host Root Port than there are UEFI-spec defined notifications, like PlatformNotify(), and PlatformPrepController(); which the platform FW can use to configure those ...
  #1 Common Clock configuration:- can be configured by platform FW if that is for pre-PCI-link configuration phase; if required for EP than  phase 1.a., has to be complete to access that as might well be in phase 2.b. itself.
  #2 Secondary PCI Express Extended Capability (PCIe Base spec 5 rev 1, chapter 7.7.3) - pre-link configuration to PCIe Root port before PCI enumeration, not intended for this ECR to acquire device-specific policy, as there are UEFI-spec defined notifications which platform FW can use to configure pre-link settings. Same as above if required for upstream port since phase 1.a has to be complete to access upstream ports which might as well be in  phase 2 itself.
  #3 Data Link Feature Extended Capability (chapter 7.7.4) - same argument as above,
  #4 Physical Layer 16.0 GT/s Extended Capability (chapter 7.7.5) - same argument as above,
  #5 Physical Layer 32.0 GT/s Extended Capability (chapter 7.7.6) - same argument as above.

If I look at phase 2.a.; it is just the PCI IO protocol installation, a SW abstraction process to identify a PCI device; and as you said platform FW can install callbacks on PCI IO protocol for certain configuration; since I want the PCI Bus driver to configure the PCI features for all the PCI devices in the system; I would not want any silent corruption to happen for the PCIe features which the EDK2 PCI Bus driver owns configuring it, and as I have highlighted earlier, a PCD is provided which is at the disposal of platform FW to dictate which PCI feature can be configured by the PCI Bus driver. If the platform FW wants to configure the PCI features at the PCI IO protocol callback, it can identify and turn off any PCI feature which the PCI Bus driver is going to configure.
Next, do I implement before phase 2.b., or after? as I informed earlier, our internal team wants it after the PCI Bus driver has processed the option ROM. But if I want to do it after phase 2.b., would not be right as the PCI Bus driver starts the PCI Bridge devices (that is by setting the PCI Command register Bus Master Enable bit). Thus, I am proposing my implementation of the RFC 2194 &amp; 2313 in the PCI Bus driver immediately after option ROM processing but before the start of PCI bridge devices. 
In short, I do not foresee any good reason where the PCI Bus driver may use the new GetDevicePolicy() interface to initialize and configure the PCIe-specific feature before the phase 1.a., or during the PCI enumeration where the PCI IO initialization happens.

&gt; From our experience not configuring certain PCI-E capabilities may affect
&gt; system&apos;s performance and sometimes system stability.
&gt; For example, changing Common Clock settings requires link retraining and
&gt; MRRS and MPLS of the Downstream devices may change. Incorrect clock
&gt; selection may degrade system performance or even make the system unusable.
&gt; 
I am totally aware of the fact for changing the common clock configuration on both sides of the link requires link retrain, and when we implement this feature for the PCI link between the Root Port and its child EP device, we can sequence this feature configuration before the other PCIe features initialization (like LTR which gets reset to default value due to DL_Down status, as per the PCI spec; so common clock configuration can happen early in the sequence before the other PCIe features).
Again, I do not see any need to split my implementation for this, to use the GetDevicePolicy() interface before the phase 1.a., above. 

&gt; As far as &quot;vendor specific workarounds&quot;, I don&apos;t think your argument is
&gt; correct. I agree that PCI bus driver loads OpROM images. However, nothing is
&gt; happening at this point. According to UEFI specification, UEFI driver should
&gt; not touch hardware in the entry point. Device programming will happen much
&gt; later, when DriverBinding-&gt;Start function is called. At this point PCI bus
&gt; driver is out of the scope(it&apos;s Start function has completed).
&gt; 
I too consider the vendor specific settings in its option ROM for a hardware workaround is a weak argument since the PCIe Base specification clearly mentions loading of the option ROM is not mandatory for the platform FW. But, the PCI Bus driver expects the image to be packaged as the PCI 2.2 Option ROM image, which can be legacy PCI option ROM or the UEFI-based option ROM image. Regardless, I am choosing the point after the PCI Option ROM processing phase 2.b., to initialize these PCI features as I don&apos;t foresee any need to configure any of the PCIe features before or during the PCI enumeration phase.

&gt; Finally, our PCI developers are still questioning the approach with changing
&gt; existing interface. Change of existing interface introduces certain level in
&gt; inconvenience (legacy code would have to be updated to switch from 
&gt; PciPlatform to PciPlatform2). This inconvenience has to be justified, by an
&gt; certain advantage that outweighs the inconvenience. It is not clear to us
&gt; what the advantage is. As I mentioned in comment 5, there a way to introduce
&gt; the functionality without changing existing protocols.
I regret the inconvenience, and I shall consult with the maintainers of the EDK2 PCI Bus driver further for their input.
With regard to justification of the advantages, for the Intel chipset alone, the PCIe ports are provided for both the north as well as south clusters of both client and server segments, and maintaining that in platform FW for the PCI complaint features without converging on code base is an overhead, would be recurring issue. The EDK2 PCI Bus driver is right place for these PCIe features and once fixed, it will be universally common solution for all the UEFI platforms.

**Comment by tianocore-issues** [comment]
## Comment 10510
Date: 2019-11-21 14:49:02 +0000
From: Felix &lt;<felixp>&gt;

Ashraf,

Thanks for such a detailed response.
You are providing arguments to demonstrate that your proposal is implementable.
I&apos;m agree that it is implementable. However, I believe that specification should define the interface, not the implementation. My proposal leaves it up to implementation whether to initialize PCI-E features earlier or later in the PCI initialization process. With your proposal early PCI-E initialization is not possible.

As you correctly mentioned, many PCI-E features are generic enough to handle them in the PCI bus driver. AMI PCI bus driver has been supporting PCI-E initialization for many years. So we do have some experience in this area. We&apos;re collecting PCI-E information as part of bus enumeration (step 1a in your comment 12) and perform PCI-E programming as a part of resource assignment (step 1b). In certain cases, it is beneficial to have complete information about capabilities and requirements of all the devices before resource assignment. From the implementation point of view it is more practical to take care about all PCI-E capabilities at the same point. That&apos;s why we&apos;re processing all supported PCI-E capabilities early.
For example, some devices may support re-sizable BARs, which means they can work faster with more resources and slower with less resources. To understand if we have enough resources to assign more to such devices we have to have full picture about resource requirements of each device in the system. Similarly, device may need more resources when SRIOV is enabled.

**Comment by tianocore-issues** [comment]
## Comment 10511
Date: 2019-11-21 17:46:24 +0000
From: @vincent-j-zimmer

responsive to https://bugzilla.tianocore.org/show_bug.cgi?id=1954#c5 

I agree w/ point 1

Why not use 
typedef struct {
  UINT8   Register;
  UINT8   Function;
  UINT8   Device;
  UINT8   Bus;
  UINT32  ExtendedRegister;
} EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL_PCI_ADDRESS;

instead of the handle, thus avoiding the temporal dependency?

And on point 2 I like the idea of just creating a new protocol for the new pci-e capability.  This will allow existing code to remain unchanged and also set a precedent for accretion of new bus technology capabilities in the future.

This compat sentiment is echoed in the last paragraph of https://bugzilla.tianocore.org/show_bug.cgi?id=1954#c11 for example.

**Comment by tianocore-issues** [comment]
## Comment 10516
Date: 2019-11-22 08:41:23 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

Alright; I shall incorporate both these proposals into this ECR.

I shall resubmit the code of PciPlatform2.h with the following changes:-
(1) Replace the EFI handle as the input of the interface GetDevicePolicy() with the 2 parameters to address the PCI device -
    (i)  PCI segment value for the target PCI device (required as Server multi-socket platforms are adopting to multi-PCI segment addressing scheme)
    (ii) PCI address value of type EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL_PCI_ADDRESS for the target PCI device

(2) New PCI Platform / Override Protocol named - EFI_PCI_E_PLATFORM_PROTOCOL and EFI_PCI_E_OVERRIDE_PROTOCOL; to support the PCIe capabilities; with the following interface members - 
    EFI_PCI_PLATFORM_GET_DEVICE_POLICY          GetDevicePolicy;
    UINT8                                       MajorVersion;
    UINT8                                       MinorVersion;

(3) Additional interface to the new PCI Platform Protocol where the consumer of this protocol shall return the PCIe features configuration values programmed to the PCI device, to the platform firmware, which can be used for other collateral. This has practical purpose for the platform to configure the PCIe Retimer data packet size in SRIS mode (reference PCI Express® Base Specification Revision 5.0 Version 1.0, chapter 4.3.11.1 Global Parameters - SRIS Link Payload Size).
    EFI_STATUS
    PlatformNotifyConfiguration (
    IN EFI_PCI_E_PLATFORM_PROTOCOL                  *This,
    IN UINT8                                        PciSegment,
    IN EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL_PCI_ADDRESS  PciAddress,
    IN EFI_PCI_PLATFORM_EXTENDED_POLICY             *PciExtPolicy
    );

Please, let me know if there is any concern in defining EFI handle as an optional input parameter of the above interfaces? The platform FW can expect this to be NULL when the consumer (PCI Bus driver) invokes the new PCI Protocol interfaces early during PCI enumeration.
Please, provide your input on how it is looking overall. 
Thanks

**Comment by tianocore-issues** [comment]
## Comment 10518
Date: 2019-11-22 09:36:41 +0000
From: Felix &lt;<felixp>&gt;

I suggest to replace PCI segment number in (1)(i) and (3) with the root bridge handle (refer to EFI_PCI_HOST_BRIDGE_RESOURCE_ALLOCATION_PROTOCOL_PREPROCESS_CONTROLLER declaration).

**Comment by tianocore-issues** [comment]
## Comment 10519
Date: 2019-11-22 09:59:23 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

(In reply to Felix from comment #16)
&gt; I suggest to replace PCI segment number in (1)(i) and (3) with the root
&gt; bridge handle (refer to
&gt; EFI_PCI_HOST_BRIDGE_RESOURCE_ALLOCATION_PROTOCOL_PREPROCESS_CONTROLLER
&gt; declaration).

I agree with this suggestion as the platform driver can use the root bridge handle to retrieve the PCI segment number. Will replace the (1)(i) and (3) with the root bridge handle.
EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL.SegmentNumber

**Comment by tianocore-issues** [comment]
## Comment 10520
Date: 2019-11-22 11:30:44 +0000
From: Felix &lt;<felixp>&gt;

Ashraf,

1. I don&apos;t like the idea of passing both PCI address and an optional handle. I think it might be confusing that two different device identification mechanisms are used. So I suggest keeping just the PCI address in GetDevicePolicy.
However, I think it might be OK to pass EFI handle (without PCI address and RB handle) to PlatformNotifyConfiguration, because the function is called once PCI-E programming is completed.
2. I suggest renaming EFI_PCI_PLATFORM_EXTENDED_POLICY to EFI_PCI_E_PLATFORM_POLICY
3. I see that all the fields of the policy structure support a special XXX_AUTO value, which is good. I think it may also be useful to add a special XXX_DO_NOT_TOUCH value to indicate that a particular PCI-E capability is managed by the platform and corresponding PCI-E registers should not be touched by the bus driver. It might be useful in dealing with non-compliant devices.

**Comment by tianocore-issues** [comment]
## Comment 10543
Date: 2019-11-25 10:11:40 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

Felix,
Regarding your #3 in comment above, to define the special XXX_DO_NOT_TOUCH value for the platform to manage particular feature won&apos;t be necessary as the PCD will be provided, at a system level, where platform firmware can use to disable that particular feature configuration in the PCI Bus driver (REF: 2194, 2313).
At a device level, the platform FW can provide a definite state as per its EFI encoding value to the PCI Bus driver through its GetDevicePolicy(); this can be the PCI-defined hardware default state itself which will result in no override of that feature for that PCI device.

**Comment by tianocore-issues** [comment]
## Comment 10545
Date: 2019-11-25 14:11:54 +0000
From: Felix &lt;<felixp>&gt;

Ashraf,

PCD token you are referring to is not part of the specification, so we cannot reply on it.

**Comment by tianocore-issues** [comment]
## Comment 10546
Date: 2019-11-26 00:00:15 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

I can update the description to clarify on the role of the PCD w.r.t. producer and consumer of this protocol; I shall also look into ways the consumer can tell the platform on what features it is ready for configuration.

**Comment by tianocore-issues** [comment]
## Comment 10614
Date: 2019-12-04 13:47:08 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

The revised PATCH V5 is sent for review to the devel@edk2.groups.io community. This change can be also reviewed in the following EDK2 repo:
https://github.com/ashrafj/edk2-staging/commit/000ce68ccbd380a3eb121e221c768f7b7ae4af50

In V5: Revised ECR to define new PCI Platform Protocol to support the
PCI Express capability features:-
EFI_PCI_EXPRESS_PLATFORM_PROTOCOL, EFI_PCI_EXPRESS_OVERRIDE_PROTOCOL.
Added new interface -&gt; NotifyDeviceConfiguration
Added new EFI encodings / data types for PCIe features:-
CPM, L1 PM substates.
Enhance the definition of the PCIe feature AtomicOp to support additional
attribute - Egress blocking of the port.

**Comment by tianocore-issues** [comment]
## Comment 10619
Date: 2019-12-05 03:58:50 +0000
From: @niruiyu

1. The new protocol name (EFI_PCI_EXPRESS_PLATFORM_PROTOCOL) looks good to me. With the new name indicating it focuses on PCIE part, carrying NO old APIs in PCI_PLATFORM protocol looks good to me.

2. The Root Bridge Handle + EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL_PCI_ADDRESS in GetDevicePolicy looks good to me. I prefer to use Handle but EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL_PCI_ADDRESS does resolve Felix&apos;s concern on the needs of early PCIE feature initialization.

3. I am not sure how the EFI_PCI_EXPRESS_PLATFORM_POLICY is passed IN and OUT from platform to PciBus driver. It&apos;s a big structure that contains all the features&apos; parameters. I guess the default value is AUTO when PciBus calls GetDevicePolicy() and platform driver is expected to change the value from AUTO to other value or leave it as is. If that&apos;s right, can you update the function header comments to describe such data flow?

4. I agree with Felix&apos;s concern that PCD should not be mentioned in the protocol header file.

5. Why do you need NotifyDeviceConfiguration?

6. The protocol header file name could be changed to PciExpressPlatform.h.

7. I&apos;d like to have a PolicySize field for EFI_PCI_EXPRESS_PLATFORM_POLICY.I understand the information may be redundant because the size can be computed from the protocol&apos;s MajorVersion and MinorVersion. But having the PolicySize may avoid potential bugs when dealing with the structure. Looking for comments on this.

**Comment by tianocore-issues** [comment]
## Comment 10620
Date: 2019-12-05 18:33:42 +0000
From: @vincent-j-zimmer

Overall great progress in evolving this design.

I agree w/ Ray&apos;s point #4 above about PCD usage.  Although we have a PCD API for dynamic ones, static ones are really an artifact of a source base and should not be mentioned in a standards-based material. 

Also, it would be valuable to update the API document along w/ these code changes. Maybe increment the version of the file name w/ a number responsive to each change and attach to the ticket?

Once folks agree on the state of things the last revision of the document file can be sent over to the UEFI Forum Platform Initialization Working Group.

**Comment by tianocore-issues** [comment]
## Comment 10632
Date: 2019-12-06 08:14:09 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

Thank you Ray and Vincent for your inputs.

Yes, I shall also provide the ECR draft along with the code changes and a version number incremental to the changes made in the document sounds good to me.

**Comment by tianocore-issues** [comment]
## Comment 10634
Date: 2019-12-06 09:23:36 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

(In reply to Ray Ni from comment #23)
&gt; 1. The new protocol name (EFI_PCI_EXPRESS_PLATFORM_PROTOCOL) looks good to
&gt; me. With the new name indicating it focuses on PCIE part, carrying NO old
&gt; APIs in PCI_PLATFORM protocol looks good to me.
&gt; 
&gt; 2. The Root Bridge Handle + EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL_PCI_ADDRESS in
&gt; GetDevicePolicy looks good to me. I prefer to use Handle but
&gt; EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL_PCI_ADDRESS does resolve Felix&apos;s concern on
&gt; the needs of early PCIE feature initialization.
&gt; 
&gt; 3. I am not sure how the EFI_PCI_EXPRESS_PLATFORM_POLICY is passed IN and
&gt; OUT from platform to PciBus driver. It&apos;s a big structure that contains all
&gt; the features&apos; parameters. I guess the default value is AUTO when PciBus
&gt; calls GetDevicePolicy() and platform driver is expected to change the value
&gt; from AUTO to other value or leave it as is. If that&apos;s right, can you update
&gt; the function header comments to describe such data flow?
The default value will be AUTO for those features which the PciBus intends to configure; and for those features it does not support it can initialize those with the value of EFI_PCI_EXPRESS_NOT_APPLICABLE when it calls the GetDevicePolicy(). This way the producer of this protocol would know for what features the PciBus driver is expecting the platform policy.
I shall update the function header comments with this details.

&gt; 
&gt; 4. I agree with Felix&apos;s concern that PCD should not be mentioned in the
&gt; protocol header file.
PCD was not mentioned before this patch. This was in response to the question on how would platform know what features would be initialized by the PciBus. Is it OK to capture the role of PCD in the ECR draft?
 
&gt; 5. Why do you need NotifyDeviceConfiguration?
This is to inform the platform on what is the final state of the features after its is configured by the PciBus. This data will be useful to platform to align any related collateral component, with the same value; like for example; the PCIe Retimer connected between the PCIe Root Port and the PCIe slot on which the EndPoint Add-In Card would be mounted, is not visible in the PCI Configuration Space, hence the platform requires the data packet size to be programmed to this PCIe Retimer components, similar to what its PCIe Root port is configured with. This notification can be used by the platform as a point to configure the PCIe Root Port&apos;s Retimer component.

&gt; 
&gt; 6. The protocol header file name could be changed to PciExpressPlatform.h.
OK.

&gt; 
&gt; 7. I&apos;d like to have a PolicySize field for EFI_PCI_EXPRESS_PLATFORM_POLICY.I
&gt; understand the information may be redundant because the size can be computed
&gt; from the protocol&apos;s MajorVersion and MinorVersion. But having the PolicySize
&gt; may avoid potential bugs when dealing with the structure. Looking for
&gt; comments on this.
The size of the EFI_PCI_EXPRESS_PLATFORM_POLICY is fixed to 128 bytes, by its definition. The definition of this protocol&apos;s major and minor version values can be used by both the consumer/producer to align on the definition of the EFI_PCI_EXPRESS_PLATFORM_POLICY.

**Comment by tianocore-issues** [comment]
## Comment 10658
Date: 2019-12-11 06:30:54 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

Created attachment 441
Revised ECR draft version 0.75: New PCI Express Protocol with EFI encoding for PCIe features

The ECR draft is revised to version 0.75 with proposal of unique definitions for the PCI Express Platform Protocol, its two interfaces, its alias, and the EFI encoding for the PCI Express features

Attachment: [BZ-441-New PCI Express Platform Protocol and EFI encodings to PCI features and its attributes.docx](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-441-New%20PCI%20Express%20Platform%20Protocol%20and%20EFI%20encodings%20to%20PCI%20features%20and%20its%20attributes.docx)

**Comment by tianocore-issues** [comment]
## Comment 10659
Date: 2019-12-11 07:45:49 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

he revised PATCH V6 is sent for review to the devel@edk2.groups.io community. This change can be also viewed in the following EDK2 repo:
https://github.com/ashrafj/edk2-staging/commit/560b6c1dd5b7eae8c49bd43e83335c7d13fef4a3

In V6: renamed the protocol definition source files, remove PCD comments,
and renamed NotifyDeviceConfiguration -&gt; NotifyDeviceState

**Comment by tianocore-issues** [comment]
## Comment 10660
Date: 2019-12-11 07:47:35 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

(In reply to Ashraf Javeed from comment #28)
The revised PATCH V6 is sent for review to the devel@edk2.groups.io
community. This change can be also viewed in the following EDK2 repo:
https://github.com/ashrafj/edk2-staging/commit/
560b6c1dd5b7eae8c49bd43e83335c7d13fef4a3

In V6: renamed the protocol definition source files, remove PCD comments,
and renamed NotifyDeviceConfiguration -&gt; NotifyDeviceState

**Comment by tianocore-issues** [comment]
## Comment 10661
Date: 2019-12-11 09:28:15 +0000
From: @niruiyu

(In reply to Ashraf Javeed from comment #26)
&gt; 
&gt; &gt; 
&gt; &gt; 4. I agree with Felix&apos;s concern that PCD should not be mentioned in the
&gt; &gt; protocol header file.
&gt; PCD was not mentioned before this patch. This was in response to the
&gt; question on how would platform know what features would be initialized by
&gt; the PciBus. Is it OK to capture the role of PCD in the ECR draft?

Platform can know which features will be initialized by PciBus by checking whether the policy value is NOT_APPLICABLE. Right?

&gt; &gt; 
&gt; &gt; 7. I&apos;d like to have a PolicySize field for EFI_PCI_EXPRESS_PLATFORM_POLICY.I
&gt; &gt; understand the information may be redundant because the size can be computed
&gt; &gt; from the protocol&apos;s MajorVersion and MinorVersion. But having the PolicySize
&gt; &gt; may avoid potential bugs when dealing with the structure. Looking for
&gt; &gt; comments on this.
&gt; The size of the EFI_PCI_EXPRESS_PLATFORM_POLICY is fixed to 128 bytes, by
&gt; its definition. The definition of this protocol&apos;s major and minor version
&gt; values can be used by both the consumer/producer to align on the definition
&gt; of the EFI_PCI_EXPRESS_PLATFORM_POLICY.

I understand PolicySize field is redundant. But considering when you increased the protocol version and enlarged the policy structure several times, how will you track the mapping relationship between the policy structure size and the protocol version?

Having the PolicySize field will produce code as below:
if (PLATFORM_POLICY.Size &gt; OFFSET_OF (PLATFORM_POLICY, FieldX)) {
  // Access PLATFORM_POLICY.FieldX
}

**Comment by tianocore-issues** [comment]
## Comment 10683
Date: 2019-12-11 22:36:11 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

(In reply to Ray Ni from comment #30)
&gt; (In reply to Ashraf Javeed from comment #26)
&gt; &gt; 
&gt; &gt; &gt; 
&gt; &gt; &gt; 4. I agree with Felix&apos;s concern that PCD should not be mentioned in the
&gt; &gt; &gt; protocol header file.
&gt; &gt; PCD was not mentioned before this patch. This was in response to the
&gt; &gt; question on how would platform know what features would be initialized by
&gt; &gt; the PciBus. Is it OK to capture the role of PCD in the ECR draft?
&gt; 
&gt; Platform can know which features will be initialized by PciBus by checking
&gt; whether the policy value is NOT_APPLICABLE. Right?

That is right! I have removed any mention to PCD in my latest changes. Please check.
&gt; 
&gt; &gt; &gt; 
&gt; &gt; &gt; 7. I&apos;d like to have a PolicySize field for EFI_PCI_EXPRESS_PLATFORM_POLICY.I
&gt; &gt; &gt; understand the information may be redundant because the size can be computed
&gt; &gt; &gt; from the protocol&apos;s MajorVersion and MinorVersion. But having the PolicySize
&gt; &gt; &gt; may avoid potential bugs when dealing with the structure. Looking for
&gt; &gt; &gt; comments on this.
&gt; &gt; The size of the EFI_PCI_EXPRESS_PLATFORM_POLICY is fixed to 128 bytes, by
&gt; &gt; its definition. The definition of this protocol&apos;s major and minor version
&gt; &gt; values can be used by both the consumer/producer to align on the definition
&gt; &gt; of the EFI_PCI_EXPRESS_PLATFORM_POLICY.
&gt; 
&gt; I understand PolicySize field is redundant. But considering when you
&gt; increased the protocol version and enlarged the policy structure several
&gt; times, how will you track the mapping relationship between the policy
&gt; structure size and the protocol version?
&gt; 
&gt; Having the PolicySize field will produce code as below:
&gt; if (PLATFORM_POLICY.Size &gt; OFFSET_OF (PLATFORM_POLICY, FieldX)) {
&gt;   // Access PLATFORM_POLICY.FieldX
&gt; }

The plan is that the size of EFI_PCI_EXPRESS_PLATFORM_POLICY would be increased, if required in future, to a block size of 128 bytes each time, along with a flip of MajorVersion. Thus, the code would change at that time to use MajorVersion as a indicator for backward compatibility to support 128B, or 256B, or 384B....

if (MajorVersion &lt; 5) {
  // 128B size of EFI_PCI_EXPRESS_PLATFORM_POLICY
  // access only up to EFI_PCI_EXPRESS_PLATFORM_POLICY.FieldX (as per its definition of protocol)

} else if (MajorVersion &lt; 10) {
  // 256B size of EFI_PCI_EXPRESS_PLATFORM_POLICY
  // access only up to EFI_PCI_EXPRESS_PLATFORM_POLICY.FieldY

} else {
  // 384B size of EFI_PCI_EXPRESS_PLATFORM_POLICY for version 10 and above
  // access only up to EFI_PCI_EXPRESS_PLATFORM_POLICY.FieldZ

}

Note that as per current version 1, the utilization is only 14B out of 128B and the rest of the bytes are for future use, so I expect very long time to complete the present 128B block before next block can be allocated (as per the protocol definition).

**Comment by tianocore-issues** [comment]
## Comment 10699
Date: 2019-12-13 01:26:36 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

(In reply to Ashraf Javeed from comment #29)
&gt; (In reply to Ashraf Javeed from comment #28)
&gt; The revised PATCH V6 is sent for review to the devel@edk2.groups.io
&gt; community. This change can be also viewed in the following EDK2 repo:
&gt; https://github.com/ashrafj/edk2-staging/commit/
&gt; 560b6c1dd5b7eae8c49bd43e83335c7d13fef4a3
&gt; 
&gt; In V6: renamed the protocol definition source files, remove PCD comments,
&gt; and renamed NotifyDeviceConfiguration -&gt; NotifyDeviceState

Kindly note that this patch V6 is align with the ECR draft version 0.75 attached here: https://bugzilla.tianocore.org/attachment.cgi?id=441

**Comment by tianocore-issues** [comment]
## Comment 10704
Date: 2019-12-13 21:55:36 +0000
From: @niruiyu

In summary, the ECR 0.75 is great with almost all feedback addressed.
Could you please review the below 3 new comments:
1. PCD.
   The PCD in your proposal is a way to let platform enable/disable certain pcie features init in PCI BUS driver.
   I think such information can be transferred to PCI BUS driver through the newly defined protocol by adding another interface:
     GetPolicy(
       IN CONST EFI_PCI_EXPRESS_PLATFORM_PROTOCOL *This,
       IN OUT EFI_PCI_EXPRESS_PLATFORM_POLICY     *PlatformPolicy
     )
   EFI_PCI_EXPRESS_PLATFORM_POLICY {
     UINTN    Size;
     BOOLEAN  MaximumPayloadSize;
     BOOLEAN  MaximumReadRequestSize;
     ...
   }

  On input (PciBus -&gt; PciePlatform), PlatformPolicy is initialized to indicate PciBus&apos;s capabilities.
  On output (PciePlatform -&gt; PciBus), PlatformPolicy is changed (set some fields to FALSE) to indicate certain feature init should be skipped in PciBus.
  &quot;Size&quot; field is set by PciBus driver as the structure size when PciBus is compiled.
  When a OLD PciBus is used with a NEW PciePlatform driver, Size set by PciBus is less than the structure size when PciePlatform is compiled. PciePlatform only changes the first &quot;Size&quot; bytes of data. It doesn&apos;t cause potential access-out-of-bound security issue.
  When a NEW PciBus is used with a OLD PciePlatform driver, Size set by PciBus is greater than the structure size when PciePlatform is compiled. It&apos;s safe for PciePlatform to access all data using the OLD structure definition when PciePlatform is compiled.

2. Change the EFI_PCI_EXPRESS_PLATFORM_POLICY in V0.75 to EFI_PCI_EXPRESS_DEVICE_POLICY because this structure is being used by GetDevicePolicy() interface and it avoids name conflict with the new structure used in GetPlatformPolicy().

3. Add a Size field to EFI_PCI_EXPRESS_DEVICE_POLICY. The reason is the same as in comment #1 adding Size to EFI_PCI_EXPRESS_PLATFORM_POLICY.


With all the comments resolved in ECR, I think we are good to have a V0.8 version (hopefully the final version).

Thanks for the effort!

**Comment by tianocore-issues** [comment]
## Comment 10707
Date: 2019-12-16 03:16:11 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

(Response to above Comment 33).

1. Regarding PCD, it also provided a way for the PciBus to initialize the PCI Express features without the new interface (GetDevicePolicy) in the legacy PCI Platform protocol to configure certain features based on the PCIe capability of a set of devices in the PCI tree. Now that, we have a new protocol, implementation specific PCD can be moved away from the PciBus; thus making this new protocol publication by the platform is the sole criteria to configure any of the PCI Express features in the platform.
I am OK with making this new protocol publication by the platform for the PciBus to react to configure the PCI Express features.
I am thinking of a new interface as below:-
EFI_STATUS
GetPolicy (
IN  EFI_PCI_EXPRESS_PLATFORM_PROTOCOL  *This,
IN  UINT8 MajorVersion,  //caller&apos;s supported version
IN  UINT8 MinorVersion,  //caller&apos;s supported version
IN  UINT8 Size,          //size of input buffer in next parameter
IN OUT  EFI_PCI_EXPRESS_PLATFORM_POLICY  *PlatformPolicy
);

typedef struct _EFI_PCI_EXPRESS_PLATFORM_POLICY {
  UINT16  Mps:1;    //set on input indicates caller support this PCIe feature. Set on output is requested feature from platform. Clear on output means platform does not require the caller to configure.
  UINT16  Mrrs:1;
  UINT16  Ro:1;
  UINT16  Ns:1;
...
  UINT16  L1PMSubstates:1;
} EFI_PCI_EXPRESS_PLATFORM_POLICY;

The PciBus can pass its supported version number which the platform can take into consideration, to return requested PCIe features. 
If the PciBus is newer than the PciePlatform, than it can know that the platform is older version based on its publication of the protocol&apos;s version data members; can reduce its version number to match with the platform.
If the PciePlatform is newer than it can come to know the PciBus version through this interface&apos;s version parameters, and it has to expect the PCIe feature list as per the protocol definition and return right PCIe feature flag in its bit field of the EFI_PCI_EXPRESS_PLATFORM_POLICY.
The PciePlatform should expect the same level of version-based PCIe feature list from the first interface call, to its subsequent device-specific interface calls through GetDevicePolicy().

2. Agree to rename the EFI_PCI_EXPRESS_PLATFORM_POLICY to EFI_PCI_EXPRESS_DEVICE_POLICY, in the GetDevicePolicy() interface. 

3. Its definition shall be revised to remove the reserved byte fields, and additional size parameter would be supplied as input parameter to this interface. The PciePlatform has to remember the PciBus version from the GetPolicy() interface, and match the size parameter to return data in output buffer of type EFI_PCI_EXPRESS_DEVICE_POLICY.

**Comment by tianocore-issues** [comment]
## Comment 10708
Date: 2019-12-16 10:09:01 +0000
From: @niruiyu

Ashraf,
I agree to put Size either in the policy structure or in the function parameter list.
But the Major/Minor version in the parameter list are unnecessary.

**Comment by tianocore-issues** [comment]
## Comment 10709
Date: 2019-12-16 12:07:12 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

Ray,
The version number of the PciBus (caller) is required for the PciePlatform to be backward compatible if it is in latest version already, compare to the caller. This shall be used to output in the EFI_PCI_EXPRESS_PLATFORM_POLICY of GetPolicy(); and also to output data in the EFI_PCI_EXPRESS_DEVICE_POLICY of GetDevicePolicy(). I don&apos;t want to rely on just the size of this input buffer, for the PciePlatform to provide its data.

**Comment by tianocore-issues** [comment]
## Comment 10727
Date: 2019-12-18 06:54:19 +0000
From: @niruiyu


Let&apos;s say DevicePolicy structures contains 3 policies added in three versions.
   DevicePolicy {
      UINT8  MaxPayloadSize;     // V1
      UINT16 MaxReadPayloadSize; // V2
      UINT8  TurboSize;          // V3
   };


Assume a OLD PciBus driver calls GetDevicePolicy with the V1 structure of DevicePolicy.

   DevicePolicy_V1 {
      UINT8 MaxPayloadSize;
   }

   GetDevicePolicy (Address, sizeof DevicePolicy_V1, &amp;DevicePolicy_V1);

The implementation of GetDevicePolicy() in the NEW PciePlatform driver can be:

   if (Size &gt; OFFSET_OF (DevicePolicy_V3, TurboSize)) {
      DevicePolicy-&gt;TurboSize = 100;
   }
   if (Size &gt; OFFSET_OF (DevicePolicy_V3, MaxReadPayloadSize)) {
      DevicePolicy-&gt;MaxReadPayloadSize = 28;
   }
   DevicePolicy-&gt;MaxPayloadSize = 16;



If Version is used instead of Size, the implementation of GetDevicePolicy() in the NEW PciePlatform driver can be:

     if (Version &gt;= 3) {
       DevicePolicy-&gt;TurboSize = 28;
     }
     if (Version &gt;= 2) {
       DevicePolicy-&gt;MaxReadPayloadSize = 18;
     }
     DevicePolicy-&gt;MaxPayloadSize = 16;


I think using Size instead of Version to control the field access is better because the code doesn&apos;t need to track which field is added by which version.

The Size usage also aligns to the cbSize field used in WinSDK:
https://docs.microsoft.com/en-us/windows/win32/api/setupapi/ns-setupapi-sp_drvinfo_data_v1_a
https://docs.microsoft.com/en-us/windows/win32/api/setupapi/ns-setupapi-sp_drvinfo_data_v2_a

If you agrees, I can help to update the V0.75 ECR to 0.8 with all comments addressed.

**Comment by tianocore-issues** [comment]
## Comment 10784
Date: 2020-01-05 02:42:03 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

Created attachment 446
Revised ECR draft version 0.80: added the new interface GetPolicy() function description

Revised ECR to define additional interface GetPolicy() in the PCI
Express Platform Protocol / PCI Express Override Protocol,
Renamed EFI_PCI_EXPRESS_PLATFORM_POLICY to EFI_PCI_EXPRESS_DEVICE_POLICY,
Removed the reserved fields from the EFI_PCI_EXPRESS_DEVICE_POLICY, added
input parameter to give the length of input buffer in GetDevicePolicy(),
and NotifyDeviceState(),
introduced new data type EFI_PCI_EXPRESS_PLATFORM_POLICY for the interface
GetPolicy().

Attachment: [BZ-446-New PCI Express Platform Protocol and EFI encodings to PCI features and its attributes.docx](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-446-New%20PCI%20Express%20Platform%20Protocol%20and%20EFI%20encodings%20to%20PCI%20features%20and%20its%20attributes.docx)

**Comment by tianocore-issues** [comment]
## Comment 10785
Date: 2020-01-05 03:02:06 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

Submitted the below Patch V7 whose code implementation aligns as per the ECR draft version 0.80 (attachment 446 [details]):-
https://github.com/ashrafj/edk2-staging/commit/cfb29efdc29fd6ae90fb2b2e304243d2dffb0c03

**Comment by tianocore-issues** [comment]
## Comment 10795
Date: 2020-01-06 22:18:26 +0000
From: @niruiyu

The protocol header file was passed review and pushed to staging repo.
The ECR content looks good to me.

**Comment by tianocore-issues** [comment]
## Comment 11127
Date: 2020-02-06 17:25:56 +0000
From: Felix &lt;<felixp>&gt;

Created attachment 456
AMI feedback based on ECR version 0.80

Ashraf,

Version 0.80 look good overall.
Please find attached updated document with AMI comments/updates.
Most(but not all) of them are editorial.
One of the comments refers to PIWG mantis ticket.
If you cannot access it, let me know.

Attachment: [BZ-456-New PCI Express Platform Protocol and EFI encodings to PCI features and its attributes.AMI.docx](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-456-New%20PCI%20Express%20Platform%20Protocol%20and%20EFI%20encodings%20to%20PCI%20features%20and%20its%20attributes.AMI.docx)

**Comment by tianocore-issues** [comment]
## Comment 11156
Date: 2020-02-06 23:37:35 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

I have completed 10 PCI Express features based on ECR 0.8 as of now, was getting ready with patches for each of the PCI Express feature to be send for code review, targeted for the edk2-staging branch UEFI_PCI_ENHANCE-2. This ECR 0.8 feedback from AMI came back late for me...can I still send the patches for code review so that it can be send to staging branch and based on further ratification of the ECR draft, the respective changes can follow based on its revised version?

**Comment by tianocore-issues** [comment]
## Comment 11158
Date: 2020-02-07 00:29:24 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

(In reply to Ashraf Javeed from comment #42)
&gt; I have completed 10 PCI Express features based on ECR 0.8 as of now, was
&gt; getting ready with patches for each of the PCI Express feature to be send
&gt; for code review, targeted for the edk2-staging branch UEFI_PCI_ENHANCE-2.
&gt; This ECR 0.8 feedback from AMI came back late for me...can I still send the
&gt; patches for code review so that it can be send to staging branch and based
&gt; on further ratification of the ECR draft, the respective changes can follow
&gt; based on its revised version?

Just to be clear, this query w.r.t. the following feature requests for which I have implemented the solution aligned with ECR 0.8 version of the new PCI Express Protocol...
For MPS, MRRS:- https://bugzilla.tianocore.org/show_bug.cgi?id=2194
For RO, NS, CTO:- https://bugzilla.tianocore.org/show_bug.cgi?id=2313
For AtomicOp, Extended Tag, LTR:- https://bugzilla.tianocore.org/show_bug.cgi?id=2499
For ASPM, CCC:- https://bugzilla.tianocore.org/show_bug.cgi?id=2500

**Comment by tianocore-issues** [comment]
## Comment 11269
Date: 2020-02-16 22:09:57 +0000
From: @niruiyu

(In reply to Felix from comment #41)
&gt; Created attachment 456 [details]
&gt; AMI feedback based on ECR version 0.80
&gt; 
&gt; Ashraf,
&gt; 
&gt; Version 0.80 look good overall.
&gt; Please find attached updated document with AMI comments/updates.
&gt; Most(but not all) of them are editorial.
&gt; One of the comments refers to PIWG mantis ticket.
&gt; If you cannot access it, let me know.

When reviewing Ashraf&apos;s implementation patch, several comments regarding to the protocol design came up:
1. GetPolicy():
   Because the platform may initialize the other features itself and doesn&apos;t expect the core to re-initialize them. GetPolicy() tells platform the core&apos;s capability of what PCIE features are supported and platform can request core to only initialize a subset of the supported features. GetPolicy() definition is good to me.

2. GetDevicePolicy():
   When a feature A is set to &quot;DISABLE&quot; by GetPolicy(). What feature specific data should GetDevicePolicy() return? Right now the ECR defines a macro *_NOT_APPLICABLE. But feature specific data like Atomic-OP and L1-PM-Substates is a structure, then what value should be filled by GetDevicePolicy() when GetPolicy() returns FALSE for Atomic-OP or L1-PM-Substates?

**Comment by tianocore-issues** [comment]
## Comment 11681
Date: 2020-03-12 09:55:38 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

Created attachment 480
Response to AMI feedback on ECR revision 0.8

Felix,
(1) All the editorial changes are accepted
(2) Concern over accepting &quot;DO NOT TOUCH&quot; option proposal due to - platform may misuse, not applicable to PCIe feature like MPS since all the PCI components of hierarchy must align to common value supported by all, AUTO and DO_NOT_TOUCH becomes 2 indefinite states for an PCIe feature when platform ought to give definite state...certain features like RO, NS and CTO can be skipped for any device, thus AUTO = DO_NOT_TOUCH for these, can define PCIe feature specific DO_NOT_TOUCH options.
(3) The AtomicOp and L1 PM sub-states policy definition shall be redefined to match other feature definitions.

Attachment: [BZ-480-New PCI Express Platform Protocol and EFI encodings to PCI features and its attributes (response to AMI).docx](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-480-New%20PCI%20Express%20Platform%20Protocol%20and%20EFI%20encodings%20to%20PCI%20features%20and%20its%20attributes%20%28response%20to%20AMI%29.docx)

**Comment by tianocore-issues** [comment]
## Comment 11692
Date: 2020-03-13 17:00:35 +0000
From: Felix &lt;<felixp>&gt;

Ashraf,

1. Your comment (3) implies that another revision of the document is coming, right?
2. If you OK with the editorial changes (your comment (1)), can you please accept them in the work document?
3. Regarding references to the EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL_PCI_ADDRESS structure (comment FP3 in the document), please replace UEFI 2.1 reference with the following statement:
&quot;EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL_PCI_ADDRESS structure is defined in the EFI_PCI_HOST_BRIDGE_RESOURCE_ALLOCATION_PROTOCOL.PreprocessController() section above.&quot;

**Comment by tianocore-issues** [comment]
## Comment 11702
Date: 2020-03-16 08:07:11 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

(In reply to Felix from comment #46)
&gt; Ashraf,
&gt; 
&gt; 1. Your comment (3) implies that another revision of the document is coming,
&gt; right?
Yes

&gt; 2. If you OK with the editorial changes (your comment (1)), can you please
&gt; accept them in the work document?
OK, it will be part of new ECR along with other editorials that are required.

&gt; 3. Regarding references to the EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL_PCI_ADDRESS
&gt; structure (comment FP3 in the document), please replace UEFI 2.1 reference
&gt; with the following statement:
&gt; &quot;EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL_PCI_ADDRESS structure is defined in the
&gt; EFI_PCI_HOST_BRIDGE_RESOURCE_ALLOCATION_PROTOCOL.PreprocessController()
&gt; section above.&quot;
I don&apos;t have access to your mantis, hence I am not sure why you are mentioning to point to PreprocessController() of the PCI Host Bridge Resource Allocation Protocol, unless the definition is being moved there.
At present, the EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL_PCI_ADDRESS structure definition is in the Root Bridge IO Protocol source header file - PciRootBridgeIo.h; and as per its location it can be defined as part of EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL.

**Comment by tianocore-issues** [comment]
## Comment 11733
Date: 2020-03-18 16:36:39 +0000
From: Felix &lt;<felixp>&gt;

&gt; I don&apos;t have access to your mantis, hence I am not sure why you are mentioning
&gt; to point to PreprocessController() of the PCI Host Bridge Resource Allocation
&gt; Protocol, unless the definition is being moved there.
&gt; At present, the EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL_PCI_ADDRESS
&gt; structure definition is in the Root Bridge IO Protocol 
&gt; source header file - PciRootBridgeIo.h; and as per its location 
&gt; it can be defined as part of EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL.

We&apos;re talking about specification update, so it does not matter in which header the structure is defined. Existing language refers to UEFI specification, which has no definition of the EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL_PCI_ADDRESS structure.
So the structure is not currently defined in any specification, which is a bug.
I&apos;m fixing it by adding definition into PreprocessController(). I&apos;ll send you a copy of my mantis ticket in a separate e-mail.

**Comment by tianocore-issues** [comment]
## Comment 11858
Date: 2020-03-26 04:25:32 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

(In reply to Felix from comment #48)
&gt; &gt; I don&apos;t have access to your mantis, hence I am not sure why you are mentioning
&gt; &gt; to point to PreprocessController() of the PCI Host Bridge Resource Allocation
&gt; &gt; Protocol, unless the definition is being moved there.
&gt; &gt; At present, the EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL_PCI_ADDRESS
&gt; &gt; structure definition is in the Root Bridge IO Protocol 
&gt; &gt; source header file - PciRootBridgeIo.h; and as per its location 
&gt; &gt; it can be defined as part of EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL.
&gt; 
&gt; We&apos;re talking about specification update, so it does not matter in which
&gt; header the structure is defined. Existing language refers to UEFI
&gt; specification, which has no definition of the
&gt; EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL_PCI_ADDRESS structure.
&gt; So the structure is not currently defined in any specification, which is a
&gt; bug.
&gt; I&apos;m fixing it by adding definition into PreprocessController(). I&apos;ll send
&gt; you a copy of my mantis ticket in a separate e-mail.

The UEFI Specification 2.8, Table 118 titled &quot;PCI Configuration Address&quot; does define the address format matching the definition of EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL_PCI_ADDRESS. Rather than defining at PreprocessController(); it could be better to define just below this Table 118 itself, as an alias to &quot;UINT64 Address&quot; the input parameter of the EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL_IO_MEM, used to define the protocol interfaces
EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL.Pci.Read()
EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL.Pci.Write()

**Comment by tianocore-issues** [comment]
## Comment 12281
Date: 2020-04-22 15:08:02 +0000
From: Ashraf Javeed &lt;<ashraf.javeed>&gt;

This PCI Express Protocol is going through further revision of removing the overheads of the EFI encoding macros of the PCIe features.

Ray demonstrated simplification of the PciBusDxe changes for the initialization of the 7 PCIe features, removed the EFI encoded macros of these features and replaced those with the actual PCIe Base specification values used with the UINT8 data type, and defined global *_AUTO and *_NOT_APPLICABLE for the feature alignment and drop scenarios, either respectively, or consequently depending on the PCIe feature.
(1) MaxPayloadSize
(2) MaxReadRequestSize
(3) RelaxedOrdering
(4) NoSnoop
(5) CompletionTimeout
(6) Ltr
(7) AtomicOp
(Ray&apos;s repo:- https://github.com/niruiyu/edk2/tree/pci/pcie2)

The review discussion are already with the devel@edk2.groups.io community. I am in the process of qualifying it further and planning to perform similar optimization to the remaining 3 PCIe features, which are already in to the edk2-staging.
(1) ExtendedTag
(2) AspmControl
(3) CommonClockConfiguration
(edk2-staging branch:- UEFI_PCI_ENHANCE-2. My repo:- https://github.com/ashrafj/edk2-staging/commits/UEFI_PCI_ENHANCE-2)

Although no plans to provide complete feature implementation support for the remaining 4 PCIe features in the PciBusDxe, its EFI encoded macros shall be revised similarly, and shall be proposed altogether in the revised PCI Express Protocol.
(1) ExtendedSynch
(2) Ptm
(3) ClockPowerManagement
(4) L1PmSubstates
